# Importing NumPy

We will primarily use NumPy for arithmetic with complex numbers as well as for linear algebra. 

To import NumPy we type the following code:

In [ ]:
import numpy as np

Arithmetic of Complex Numbers

In Python the imaginary unit $i$ is instead represented as $j$, and must always be written with a coefficient. So in Python, we must write $1j$ instead of $j$, and $-1j$ instead of just $-j$. Let's check that $j^2=-1$.

In [1]:
(-1j)*(-1j)

(-1+0j)

As we can see, the output is:

$-1 + 0j = -1.$


Now, let's define two complex numbers

$z = 3 + 2j$

$w = -2 -5j$

In [2]:
z = 3 + 2j
w = -2 - 5j

If we wanted to print $z$ and $w$, we could simply type them into a Jupyter cell as follows (followed by shift+enter):

In [3]:
z

(3+2j)

In [4]:
w

(-2-5j)

We could also type the print() command (followed by shift+enter):

In [5]:
print(z)

(3+2j)


In [6]:
print(w)

(-2-5j)


If we wanted to label what we are printing we can do the following:

In [7]:
print("z=", z)

z= (3+2j)


In [8]:
print("the complex number w is:", w)

the complex number w is: (-2-5j)


the complex number $w$ is: $(-2-5j)$

We can print the real and imaginary parts of complex numbers as follows for $z = 3+2i$:

In [11]:
print('z=', z)
print('Real(z)=', np.real(z))
print('Imag(Z)=', np.imag(z))

z= (3+2j)
Real(z)= 3.0
Imag(Z)= 2.0


In [13]:
print('w=', w)
print('Real(w)=', np.real(w))
print('Imag(w)=', np.imag(w))

w= (-2-5j)
Real(w)= -2.0
Imag(w)= -5.0


Complex Conjugates

We can compute the $\bf{complex}$ $\bf{conjugates}$ $z^*$ and $w^*$, which just negates the imaginary part of the complex numbers as follows:

In [12]:
np.conj(z)

np.complex128(3-2j)

In [14]:
np.conj(w)

np.complex128(-2+5j)

As we can see, complex conjugation negates the imaginary part of the complex number.


# **Norms/Absolute Values**

We can compute the $\bf{absolute}$ $\bf{value}$ of the complex numbers:


$
\|z\|=\sqrt{zz^*}=\sqrt{{\vert{z}\vert}^2},
\|w\|=\sqrt{ww^*}=\sqrt{{\vert{w}\vert}^2},
$

In [15]:
print('||z||=', np.abs(z))

||z||= 3.6055512754639896


In [16]:
print('||w||=', np.abs(w))

||w||= 5.385164807134504


$||w||= 5.385164807134504$

Now, there are many different arithmetic operations we can perform on complex numbers. Addition, subtraction, multiplication, and division are all defined for complex numbers and can be computer using NumPy. As an example, let's compute the following three values,


*   z + w
*   z - w
*   w - z





In [17]:
z+w

(1-3j)

In [18]:
z-w

(5+7j)

In [19]:
w-z

(-5-7j)

Addition and subtraction of complex numbers is done componentwise, meaning the real parts and the complex parts are added or subtracted individually and do not effect each other. So, for example,

$z + w = (3 + 2j) + (-2 - 5j) = (3 - 2) + (2 - 5) j = 1 - 3j$

We can also multiply complex numbers and divide them as well. This is also handled by NumPy. Let's compute the following three values,



*   $z.w$
*   $\frac{z}{w}$
*   $\frac{w}{z}$



In [20]:
z*w

(4-19j)

In [21]:
z/w

(-0.5517241379310345+0.37931034482758624j)

In [22]:
w/z

(-1.2307692307692308-0.8461538461538463j)


# Vectors

# Introduction

Vectors will be fundamental in our study of quantum computing. The most basic unit of computation in a quantum computer is a *qubit*, which can be represented as a -dimensional complex vector of length one. So understanding vectors will be foundational and necessary for most of what we will be doing in this book. Vectors can be thought of in many ways, one of the most basic is simply as an array of numbers, which we will often represent as a column of numbers called **column vectors**, but in some cases we will also need **row vectors**:

Column Vector: $\begin{pmatrix}
    a_1 \\
    a_2 \\
    \vdots\\ 
    a_n
  \end{pmatrix}$, 
  
  Row Vector: $\begin{pmatrix} a_1, & a_2, & \ldots, & a_n \end{pmatrix}$


# Row Vectors, Column Vectors, and Bra-Ket Notation

We can create a column vector and a row vector in Python:

In [ ]:
# Create a vector as a row
row_vector = np.array([2-1j, 7j, -3])

# Create a vector as a column
column_vector = np.array([[2+1j],
                          [-5],
                          [2j]])

Row vectors in quantum mechanics are also called **bra-vectors**, and are denoted as follows:


$⟨A|=\begin{pmatrix} a_1, & a_2, & \ldots, & a_n \end{pmatrix}$

Column vectors are also called **ket-vectors** in quantum mechanics denoted as follows:

$|B⟩=\begin{pmatrix}
    b_1 \\
    b_2 \\
    \vdots\\ 
    b_n
  \end{pmatrix}$

  In general, if we have a column vector, i.e. a ket-vector:

  $|A⟩=\begin{pmatrix}
    a_1 \\
    a_2 \\
    \vdots\\ 
    a_n
  \end{pmatrix}$

  the corresponding bra-vector:

  $⟨A|=\begin{pmatrix} a_1^*, & a_2^*, & \ldots, & a_n^* \end{pmatrix}$


  is the complex-conjugate transpose of the ket-vector , and vice-versa. As an example, if we have the following ket-vector:

  $|A⟩=\begin{pmatrix}
    2+i \\
    1-3i \\
  \end{pmatrix}$

  then the corresponding bra-vector is:

  $⟨A|=\begin{pmatrix} 2-i, & 1+3i \end{pmatrix}$

The notation $⟨A|$ and $|A⟩$, for bra- and ket-vectors respectively, is a reference to *inner products* denoted by brackets and we will discuss it more in the next section when we define inner product. As another example, we might want to have a pair of $2$-dimensional complex column vectors (meaning the entries are complex numbers):

$|A⟩=\begin{pmatrix}
    2-i \\
    5 \\
  \end{pmatrix}$

  $|B⟩=\begin{pmatrix}
    7 \\
    3i \\
  \end{pmatrix}$

  Remember, $i^2=-1$
 and so $i=\sqrt{-1}$ is imaginary and complex numbers are always of the form . The corresponding bra-vectors would then be:

 $⟨A|=\begin{pmatrix} 2+i, & 5 \end{pmatrix}$

 $⟨B|=\begin{pmatrix} 7, & -3i \end{pmatrix}$

The $2$-dimensional vectors we have listed above are not quite the kind of vectors we will be using most often in quantum computing, but until we introduce the norm of a vector and explain what unit vectors are, let us focus on the basic operations we can perform on these vectors. We can always add two vector of the same dimension:

$|A⟩ + |B⟩=\begin{pmatrix}
    2-i \\
    5 \\
  \end{pmatrix}+ \begin{pmatrix}
    7 \\
    3i \\
  \end{pmatrix} = \begin{pmatrix}
    (2-i)+7 \\
    5 +3i \\
  \end{pmatrix} = \begin{pmatrix}
    9-i \\
    5+3i \\
  \end{pmatrix}$

  Task 1: add these two kets in Python:

In [23]:
### TO DO ###

ket_A = np.array([[2-1j], [5+0j]])
ket_B = np.array([[7+0j], [0+3j]])

res_ket = ket_A + ket_B
print(res_ket)

[[9.-1.j]
 [5.+3.j]]


We can multiply bra-vectors and ket-vectors by scalars:

$3|A⟩=3\begin{pmatrix}
    2-i \\
    5 \\
  \end{pmatrix}=\begin{pmatrix}
    6-3i \\
    15 \\
  \end{pmatrix}$

In [24]:
print(3*ket_A)

[[ 6.-3.j]
 [15.+0.j]]


Similarly for row vectors (bra-vectors):

$5⟨B|=5\begin{pmatrix} 7, & 3i \end{pmatrix}=\begin{pmatrix} 35, & 15i \end{pmatrix}$

In [25]:
bra_B = np.array([[7, 3j]])
print(5*bra_B)

[[35. +0.j  0.+15.j]]


**Exercises**




1.   Define the following ket-vector as an array in Python:

  $|$ψ$⟩=\begin{pmatrix} 2, & 3i \end{pmatrix}$
2.   Define $|$ψ$⟩$ from the previous problem as a (column) matrix and compute its Hermitian conjugate.


3.   Define two ket vectors (column vectors) in Python:

$|A⟩=\begin{pmatrix}
    2 \\
    5-5j \\
  \end{pmatrix}$, $|B⟩=\begin{pmatrix}
    2 + 3j \\
    -3j \\
  \end{pmatrix}$
4.   Compute $|A⟩ + |B⟩ $.

5.   Compute $3|A⟩$.









In [26]:
# 1
ket_1 = np.array([[2 + 0j], [0 + 3j]])

In [29]:
# 2
matrix_2 = np.matrix(ket_1)
matrix_2.H

matrix([[2.-0.j, 0.-3.j]])

In [30]:
# 3 
ket_3_A = np.array([[2 + 0j], [5 - 5j]])
ket_3_B = np.array([[2 + 3j], [0 - 3j]])

In [31]:
# 4
ket_3_A + ket_3_B

array([[4.+3.j],
       [5.-8.j]])

In [32]:
3 * ket_3_A

array([[ 6. +0.j],
       [15.-15.j]])